doc/dev/mountpoints.md

```diff
@@ -0,0 +1,179 @@
+# Proposal: New mountpoint config structure
+
+A mountpoint is defined by adding keys below `system:/elekta/mountpoints/<mountpoint>`, where `<mountpoint>` is the key name for the parent key of the mountpoint with slashes properly escaped.
+For example for the mountpoint `user:/mymountpoint` you have to add keys below `system:/elekta/mountpoints/user:\/mymountpoint`.
+
+Every mountpoint consists of exactly three parts:
+
+1. A list of plugins required for the mountpoint.
+2. A single _backend plugin_ chosen from the list of plugins.
+3. The _mountpoint definition_ specific to the backend plugin.
+
+The list of plugins is defined as the array `system:/elektra/mountpoints/<mountpoint>/plugins/#`.
+Specifically, `system:/elektra/mountpoints/<mountpoint>/plugins/#/name` contains the name of the plugin and `system:/elektra/mountpoints/<mountpoint>/plugins/#/config` the configuration keyset passed to the plugin.
+
+The single _backend plugin_ is the only one that the `libelektra-kdb` library communicates with.
+It is responsible for calling other plugins when needed.
+A _backend contract_ exists between this plugin and `libelektra-kdb`.
+The backend plugin alone is responsible for upholding this contract.
+
+The backend plugin must be one of the ones defined in `system:/elektra/plugins/#`.
+The tell `libelekta-kdb` which of the plugins to use as the _backend plugin_, you set `system:/elektra/mountpoints/<mountpoint>/backend` to its array index.
```

Comment:
Why not simply say `#0` is the backend plugin?

Reply:
Because I prefer an explicit self-describing config over implicit conventions that need to be remembered. Especially, if there is no technical reason one way or the other.

Reply:
I agree that Elektra overused implicit conventions. In this case, however, there are technical reasons:

- at least two keys are needed, so a single kdb set will lead to a broken (unfinished) mountpoint
- the reference to the backend could be wrong (e.g. out of bounds), especially when reordering the array (we don't have any validation when backends are constructed, and even if there were, current plugins would not be able to validate this)

Reply:
> at least two keys are needed, so a single `kdb set` will lead to a broken (unfinished) mountpoint

This is a general problem with `kdb set` and not a problem with this specification. There are many other cases where `kdb set` is not usable.

> the reference to the backend could be wrong (e.g. out of bounds),

So what? The name given in the array could also be wrong (i.e. plugin doesn't exist). It is another possible error, yes. But it's not like it's the only possible error.

> especially when reordering the array 

Yes, if an element before the backend plugin is removed the index would change.

However, this may not be a problem. I'm actually thinking of allowing arbitrary names in `system:/elektra/mountpoints/<mountpoint>/plugins` instead of enforcing an array. I don't actually see any benefit in using array. We don't need the order, we just need to have some way of referencing a plugin.

> we don't have any validation when backends are constructed, and even if there were, current plugins would not be able to validate this

I'm not sure what exactly you mean here.

`kdb mount` should check the configuration and not create broken ones. You could use `kdb set` to create a broken config, but you could also manually edit the file where the config is stored.

`kdbOpen()` checks the config and fails, if it is broken. That means once it's broken, you need to manually edit the config file. A simple way around that would be to add an option (e.g. via the contract). If the option is set `kdbOpen()` would only read the hardcoded and unchangable backends (this includes `system:/elektra/mountpoints`). This "recovery" mode would make it possible to always fix a broken mountpoint.

It would also be possible to create a plugin (or a part of `libelektra-kdb` itself) that validates `system:/elektra/mountpoints` during every `kdbSet`. The "recovery" mode would still be useful, in case a mountpoint is broken because of a missing plugin.

Reply:
Well designed config can be changed with setting individual key/values without producing intermediate states where something is broken.

I would prefer if this would be true for the backend configuration. With the small exception of `system:/elektra/mountpoints/<mountpoint>/backend` this is already the case. I even have to admit that it is much more elegant than it was before.

Reply:
If I want to add a plugin with a config KeySet, I already need to set more than one key at at time.

IMO for a "well designed config" it is more important, that it is understandable and somewhat self-documenting. Even if that means you sometimes have to consider multiple settings as one unit.

In this particular case, I think we can reach a compromise. We don't make `system:/elektra/mountpoints/<mountpoint>/plugins` and array and we remove `system:/elektra/mountpoints/<mountpoint>/backend`. Instead `system:/elektra/mountpoints/<mountpoint>/plugins/<name>` can be used to define plugins, where `<name>` is any name other than `backend`. The key ``system:/elektra/mountpoints/<mountpoint>/plugins/backend` is used to define the designated backend plugin.

So the example would look like this (together with the suggestion to move the mountpoint definition part from below):

```
# List of plugins with their config
system:/elektra/mountpoints/\/hosts/plugins/resolver/name (="resolver_fm_hpu_b")
system:/elektra/mountpoints/\/hosts/plugins/myglob/name (="glob")
system:/elektra/mountpoints/\/hosts/plugins/myglob/config/set/#0
system:/elektra/mountpoints/\/hosts/plugins/myglob/config/set/#1
system:/elektra/mountpoints/\/hosts/plugins/myglob/config/set/#2
system:/elektra/mountpoints/\/hosts/plugins/myglob/config/set/#3
system:/elektra/mountpoints/\/hosts/plugins/myglob/config/set/#4/flags
system:/elektra/mountpoints/\/hosts/plugins/store/name (="hosts")
system:/elektra/mountpoints/\/hosts/plugins/sync/name (="sync")
system:/elektra/mountpoints/\/hosts/plugins/error/name (="error")
system:/elektra/mountpoints/\/hosts/plugins/net/name (="network")

# Define backend plugin
system:/elektra/mountpoints/\/hosts/plugins/backend/name (="backend")

# Mountpoint definition for backend plugin
system:/elektra/mountpoints/\/hosts/definition/path (="myhosts")

system:/elektra/mountpoints/\/hosts/definition/positions/get/resolver = (="resolver")
system:/elektra/mountpoints/\/hosts/definition/positions/get/storage = (="store")
system:/elektra/mountpoints/\/hosts/definition/positions/get/poststorage/#0 (="myglob")

system:/elektra/mountpoints/\/hosts/definition/positions/set/resolver = (="resolver")
system:/elektra/mountpoints/\/hosts/definition/positions/set/prestorage/#0 = (="myglob")
system:/elektra/mountpoints/\/hosts/definition/positions/set/prestorage/#1 = (="error")
system:/elektra/mountpoints/\/hosts/definition/positions/set/prestorage/#2 = (="net")
system:/elektra/mountpoints/\/hosts/definition/positions/set/storage = (="store")
system:/elektra/mountpoints/\/hosts/definition/positions/set/precommit/#0 = (="sync")
system:/elektra/mountpoints/\/hosts/definition/positions/set/commit = (="resolver")
system:/elektra/mountpoints/\/hosts/definition/positions/set/rollback (="resolver")
```

This is a bit more complicated to parse, but I think it is generally easier to understand/write, because you can use custom names to refer to plugins instead of having to remember an index.

Reply:
Discussion result: We will use the config with names as proposed in the comment above. Tools like `kdb mount` will automatically generate names from the plugin name or the plugin contract. `kdb mount` will only support the default backend, using other backend plugins must be done manually.

Additional results: The plugin contract should be returned by a separate function, not a special `elektra<Plugin>Get` call. Probably it will replace the current plugin factory function. The plugin contract will contain the type of plugin (normal, backend, global(?), etc.) and maybe also the set of backend plugins/positioning schemes a plugin is compatible with.

Reply:
> Tools like kdb mount will automatically generate names from the plugin name or the plugin contract.

I think we need to clarify this beforehand. Otherwise we end up with a semi-array (`#0`, `#1`, ... but with one key that is called `backend`). I don't see a good way to generate names without index if the same plugin is used several times.

I still think `#0` for the backend plugin is the easiest choice: for both `kdb mount` and for manual mountpoints (which get practically irrelevant once we have a `kdb mount` implementation in C).

Even if you argue that the ordering does not have a meaning, the backend plugin must be the one which gets loaded first anyway, so `#0` makes sense, even if the other index would be arbitrary. I don't see that the index are arbitrary: the user specify the plugins in an ordered way (the arguments in `kdb mount`) and there are cases where the loading or execution of plugin needs order, so array seems to me like a perfect fit.

> -B

Yes, I like the idea to actually have non-file based plugins but I see it the same way: I would leave it for 1.0 now as-is. I don't think we need anything but the file backend for 1.0.

Reply:
> I think we need to clarify this beforehand. Otherwise we end up with a semi-array (`#0`, `#1`, ... but with one key that is called `backend`).

We already this clarified this in our last discussion. There is no need for an array here. We don't need an order, we just need a name to reference the plugin instance. `kdb mount` will only support a single instance per plugin and will simply use the plugin name. 

> Even if you argue that the ordering does not have a meaning, the backend plugin must be the one which gets loaded first anyway

Why? 

>  there are cases where the loading or execution of plugin needs order, so array seems to me like a perfect fit.

What use case would need a specific order of `kdbOpen` calls?

The order of `kdbGet`/`kdbSet` is of course important, but that is defined in the `definition` part (that is why e.g. `get/poststorage/#` is an array).

> I don't think we need anything but the file backend for 1.0.

But we should still plan for them, to minimize breaking changes post-1.0

Reply:
> We already this clarified this in our last discussion.

We didn't speak about tooling. The discussion also didn't have a consent, you proposed something and in the next post you said that this proposal is the result of the discussion :wink:  We definitely should clarify how the tooling will work because to require the user invent an name for every plugin in a backend would be insane. We had something like this for the backend names, which was not a good idea. With arrays we wouldn't have this problem at all.

> What use case would need a specific order of kdbOpen calls?

Of course you try to avoid it, but kdbOpen calls sometimes have side effects, e.g. have some global initialization. E.g. you would like to do: `kdb mount .. .. java_plugin_1 jvm/start=1 java_plugin_2` and then you obviously want that java_plugin_1 gets loaded first.

Reply:
> We didn't speak about tooling. The discussion also didn't have a consent, you proposed something and in the next post you said that this proposal is the result of the discussion wink 

We **did** _speak_ about tooling and agreed on exactly what I marked as "Discussion result" in our speparate meeting at the end of July. See also https://github.com/ElektraInitiative/libelektra/pull/3693#issuecomment-886646409

> Of course you try to avoid it, but kdbOpen calls sometimes have side effects, e.g. have some global initialization. E.g. you would like to do: kdb mount .. .. java_plugin_1 jvm/start=1 java_plugin_2 and then you obviously want that java_plugin_1 gets loaded first.

Because of the way plugins are loaded, this is exactly what you _don't_ want to do. As described in `doc/dev/mountpoints.md` (around line 40), expensive initialization (like creating a JVM instance) should _not_ happen in `elektra<Plugin>Open`. Instead it should be done lazily during the first `elektra<Plugin>Open` (*). This is because `elektra<Plugin>Open` is always called for all plugins during `kdbOpen`., even those not

Reply:
> We did speak about tooling

I cannot remember that we found any solution for generating the names. We were only discussing that this problem exists.

> this is exactly what you don't want to do.

We might not want it but we are required to do it because of the libs the plugins use.

You did not respond to the issue that for kdbGet/kdbSet there is also an order, e.g. between plugins that compress and encrypt. I do not see a benefit to make a random order (what it more or less would be if we have a name generation algorithm) or always enforce order (via plugin contracts). I would keep it simple and use the order `kdb mount` gives us.